### DISCUSSION TITLE
Efficiency with Ryzen 5900x, 128GB ram and 2TB ‎NVMe (FireCuda 520)

### DISCUSSION THREAD

**Discussion by adamsky-01** [madMAx43v3r/chia-plotter#150] [General]
Did somebody test this plotter with Ryzen 5900x? What could be the expected time per plot with the below configuration: 
- Ryzen 5900x 
- 128 GB 4 x 32 GB DDR4 3600 MHz
- 2TB ‎NVMe Gen4 (FireCuda 520)

I'm just considering the ram upgrade to128GB.

UPDATE:
RAM Disk works pretty well for me. I'm running windows v0.0.6 version of plotter (256 bucket, 24 threads) with 112GB assigned to RAM Disk.
Total plot creation time: 28 min

**Comment by casa10** [comment]
Same question with 64GB of RAM (3600MHz)

**Comment by delucca** [comment]
I have a Ryzen 5900x with 32GB of RAM and a 970EVO

I'm taking 1h to plot k32

Since this plotter optimizes your plotting speed, it does not make sense to run parallel plots. So, I don't see any benefits in large RAMs, unless you use RAM as your tmp2 dir. The same applies to large NVMe

**Reply by adamsky-01** [reply]
> I have a Ryzen 5900x with 32GB of RAM and a 970EVO
> 
> I'm taking 1h to plot k32
> 
> Since this plotter optimizes your plotting speed, it does not make sense to run parallel plots. So, I don't see any benefits in large RAMs, unless you use RAM as your tmp2 dir. The same applies to large NVMe

Thanks for sharing this info. I'm actually thinking about 110GB RAM disk to reduce the NVMe wear out.  The other question is will the remaining 128GB-110GB ram memory be enough to utilize all threads of 5900X for plotting?

**Reply by delucca** [reply]
I don't think 110GB would be enought to work as a tmp2 disk, since it needs 110GiB (notes the `i`). Problably you would need ~119GB of your ram as the tmp2 disk

In any case, even if you use 119GB as tmp2 disk, you would still have 9gb

If you're running on Linux that should be enough to keep remaining processes running.

**Reply by casa10** [reply]
> 
> 
> I have a Ryzen 5900x with 32GB of RAM and a 970EVO
> 
> I'm taking 1h to plot k32
> 
> Since this plotter optimizes your plotting speed, it does not make sense to run parallel plots. So, I don't see any benefits in large RAMs, unless you use RAM as your tmp2 dir. The same applies to large NVMe

thank you for your answer.. 
1 ) You use which operating system ? (Ubuntu, Centos, windows, etc.) and which version of the OS too please ?
2) how do you configure to avoid to be over the max DDR RAM (in your case 32Gb DDR RAM)

**Reply by kenfookchoong** [reply]
Good day, may i ask
excluding the ram disk usage

the plotting will still using 3.4GB ram for plotting ?

**Reply by delucca** [reply]
> > I have a Ryzen 5900x with 32GB of RAM and a 970EVO
> > I'm taking 1h to plot k32
> > Since this plotter optimizes your plotting speed, it does not make sense to run parallel plots. So, I don't see any benefits in large RAMs, unless you use RAM as your tmp2 dir. The same applies to large NVMe
> 
> thank you for your answer..
> 1 ) You use which operating system ? (Ubuntu, Centos, windows, etc.) and which version of the OS too please ?
> 2) how do you configure to avoid to be over the max DDR RAM (in your case 32Gb DDR RAM)

1) Linux Manjaro (Arch), but the plot speed should be agnostic regarding the OS, since it is compiled in C
2) I set the RAM usage to 26gb, it more than enough!

**Reply by delucca** [reply]
> Good day, may i ask
> excluding the ram disk usage
> 
> the plotting will still using 3.4GB ram for plotting ?

I think it depends on the number of cores you have

**Reply by madMAx43v3r** [reply]
`tmp2` needs exactly 104 GiB, 128 GiB RAM is enough for about 12 threads, so it could just work out by a hair.

**Reply by madMAx43v3r** [reply]
> I'm taking 1h to plot k32

You're def limited by that EVO, that CPU should do 30min plots easy

**Reply by madMAx43v3r** [reply]
Note: RAM is measured in GiB, not GB

**Reply by adamsky-01** [reply]
> `tmp2` needs exactly 104 GiB, 128 GiB RAM is enough for about 12 threads, so it could just work out by a hair.

Thank you for the clarification. Most of the consumer grade motherboard is limited to 128 GiB of ram. This means only 12 threads can be utilized, so saving NVMe with 104 GiB ram disk and max 12 CPU threads or probably faster plotting with all threads but no ram disk benefit. 
Will this limit of 12 threads (with 104 GiB ram disk and 128 GiB memory) significantly increase the plotting time? What time could be expected in early mentioned configuration with Ryzen 5900x (12 threads used only)?

**Reply by casa10** [reply]
i don't understand all.. Sorry..not clear from beginning for me.
- you can use your ddr RAM like temp2 directory. In that case, 104GiB is necessary. If you use 12 threads, you need less than 4GiB of RAM so 128GiB it's ok because you consider for example OS take 4-6Gib of RAM.
- if you have only 64 GiB of RAM (or less 32GiB etc.), can you configure tmp2 directory in this application ? Does she will write step by step from ddr RAM to SSD (or HDD) for the 104 GiB and used in same time the 220 GiB of space in the SSD ? In this case of use, do you write less in the SSD in total than using swar manager or plot manager where 1.8 TB of the SSD is using ?

**Reply by marctanw90** [reply]
weird. i threw 24 cores on my 5900x but its CPU usage is fluctuates a lot and averages around 33%

**Reply by casa10** [reply]
i tried with 128GiB of RAM tmp2 (ddr ram) and used tmp1 SSD.
40 min to make 1 plot with 12 threads.
CPU : 72°C max
EDC : 100% for all phase P1, 86-88% phase P2, 79-93% phase P3
Consumption : 102-115W for P1,90W for P2 , max 105W for P3 
PPT : max 100%  but often between 90-95%
I don't know if for CPU it's good to work like that often and 24/24hours.

**Reply by adamsky-01** [reply]
> CPU : 72°C max

You are absolutely fine with this temp. on your CPU even 24/7 over years.

**Comment by thiagottjv** [comment]
Hi, I have a 5900x, 32Gb of RAM and 2 970 evo 2Tb (one for each temp destination).
I`m doing one plot with around 1750s (30min). Each time a plot ends this time increase.
With that said, I`m using only around 16 to 17Gb of ram using all 24 threads. I`ll try to decrease the buckets to increase the ram usage and hopefully decrease the time.

**Reply by adamsky-01** [reply]
> Hi, I have a 5900x, 32Gb of RAM and 2 970 evo 2Tb (one for each temp destination).
> I`m doing one plot with around 1750s (30min). Each time a plot ends this time increase. With that said, I`m using only around 16 to 17Gb of ram using all 24 threads. I`ll try to decrease the buckets to increase the ram usage and hopefully decrease the time.

THANKS! for sharing this. I'm just waiting for 128GiB RAM to make RAMdisk and use it for tmp2 directory.

**Reply by thiagottjv** [reply]
Please post your results.
Just to append my comments: that time was with one plot.
Plotting in a row is taking 1850s for the plot and 750s for the transfer witch means around 46 plots a day.

**Reply by adamsky-01** [reply]
I run a quick test under Windows 10, no other tasks just single plot using madMAx43v3r/chia-plotter windows build v0.0.4
chia_plot.exe -n 1 -r 24 (5900X) -t (Samsung 970 EVO 500GB NVMe) -2 (FireCuda 520 2TB ‎NVMe Gen4 ) 

Phase 1 took 1168.58 sec
Phase 2 took 525.656 sec
Phase 3 took 694.047 sec
Phase 4 took 142.096 sec
Total plot creation time was 2530.42 sec (42 min)

It could be probably faster under the Linux. I'll update these results with the RAM disk and under Ubuntu when I'll receive the 128GiB RAM kit.

**Reply by adamsky-01** [reply]
Changing buckets from default 128 to 256 helped me to save 5 minutes on plot. Setting as above, still on Windows 10.

Phase 1 took 881.111 sec
Phase 2 took 510.704 sec
Phase 3 took 726.681 sec
Phase 4 took 104.45 sec
Total plot creation time was 2223.04 sec (37 min)

**Reply by thiagottjv** [reply]
Yes, seams that the way to go is increase the bucket size... With 128 first plot time was at around 1750s, with 64 it wont finish because I didn't had enough ram, but the phase 1 timings were higher.
With 256 I got 1630s, with 512, 1580s! So I went down almost 3min!
I`ll try to get higher.
The next test will be doing a raid 1 with my two nvmes and using the same folder for temp1 and 2.

**Reply by adamsky-01** [reply]
I've finished test with 64 buckets but the result was very similar to 128 buckets. Below results for the direct comparison. 
OS: Windows 10
Plotter: chia-plotter windows build v0.0.4
CPU: Ryzen 5900X
RAM: 48GiB 3200MHz CL16
tmpdir: Samsung 970 EVO 500GB NVMe
tmpdir2:  FireCuda 520 2TB ‎NVMe Gen4

64 buckets
Phase 1 took 1182.35 sec
Phase 2 took 520.179 sec
Phase 3 took 702.441 sec
Phase 4 took 96.3558 sec
Total plot creation time was 2501.42 sec (42 min)

128 buckets
Phase 1 took 1156.46 sec
Phase 2 took 513.086 sec
Phase 3 took 736.826 sec
Phase 4 took 103.345 sec
Total plot creation time was 2509.79 sec (42 min)

256 buckets
Phase 1 took 881.111 sec
Phase 2 took 510.704 sec
Phase 3 took 726.681 sec
Phase 4 took 104.45 sec
Total plot creation time was 2223.04 sec (37 min)

**Reply by adamsky-01** [reply]
Update:
RAM Disk works pretty well for me. I'm running windows v0.0.6 version of plotter (256 bucket, 24 threads) with 112GB assigned to RAM Disk.
Total plot creation time: 28 min

**Comment by jacek900** [comment]
For your configuration - with ram disk - the total time should be around 29 minutes for plot total. Mine is 1788 seconds but my memory is slower.
I am going to test RAID0 for temp1 to see if it speeds up phase 2 as my cpu seems to be a bit bored during that one.

**Reply by adamsky-01** [reply]
Dzięki za info Jacek, but my current plot time (42 minutes) seems a bit slow. I just hope it's Windows and the early build of madMAx43v3r plotter for this system not any hardware bottleneck.

**Reply by jacek900** [reply]
This is my last plot. The time varies slightly from plot to plot but its slightly more than 2 plots per hour average.
Recommend switching to Ubuntu or other system.
Comparing times with the author - my phase 2 is a bit slow. I can also see that on htop.

Process ID: 5560
Number of Threads: 24
Number of Buckets: 2^7 (128)
Pool Public Key:   
Farmer Public Key: 
Working Directory:   /mnt/ssdtemp2/
Working Directory 2: /mnt/ramdisk/
Plot Name: plot-k32-2021-06-13-12-16-79a13c1d7470e6d244ffa050af1f14f725b1ea9c6f04571a4d61f1e598ad0509
[P1] Table 1 took 13.0594 sec
[P1] Table 2 took 105.145 sec, found 4294966611 matches
[P1] Table 3 took 150.852 sec, found 4294960144 matches
Copy to /mnt/chiaroboczy/plot-k32-2021-06-13-11-46-fb90d61e3a2a84353c78f2f3dc8eb9af15e473d3b697b3d370d76677f01a0683.plot finished, took 365.356 sec, 284.09 MB/s avg.
[P1] Table 4 took 183.818 sec, found 4295045825 matches
[P1] Table 5 took 175.408 sec, found 4294928086 matches
[P1] Table 6 took 166.458 sec, found 4294822646 matches
[P1] Table 7 took 114.103 sec, found 4294716442 matches
Phase 1 took 908.848 sec
[P2] max_table_size = 4295045825
[P2] Table 7 scan took 5.98642 sec
[P2] Table 7 rewrite took 16.1319 sec, dropped 0 entries (0 %)
[P2] Table 6 scan took 28.5384 sec
[P2] Table 6 rewrite took 48.9795 sec, dropped 581304092 entries (13.535 %)
[P2] Table 5 scan took 30.8724 sec
[P2] Table 5 rewrite took 50.5591 sec, dropped 761984501 entries (17.7415 %)
[P2] Table 4 scan took 32.953 sec
[P2] Table 4 rewrite took 46.0189 sec, dropped 828912268 entries (19.2993 %)
[P2] Table 3 scan took 32.1928 sec
[P2] Table 3 rewrite took 45.8063 sec, dropped 855096678 entries (19.9093 %)
[P2] Table 2 scan took 35.2611 sec
[P2] Table 2 rewrite took 45.5827 sec, dropped 865581021 entries (20.1534 %)
Phase 2 took 428.759 sec
Wrote plot header with 268 bytes
[P3-1] Table 2 took 38.8394 sec, wrote 3429385590 right entries
[P3-2] Table 2 took 27.4814 sec, wrote 3429385590 left entries, 3429385590 final
[P3-1] Table 3 took 35.8685 sec, wrote 3439863466 right entries
[P3-2] Table 3 took 29.1129 sec, wrote 3439863466 left entries, 3439863466 final
[P3-1] Table 4 took 38.4285 sec, wrote 3466133557 right entries
[P3-2] Table 4 took 29.1823 sec, wrote 3466133557 left entries, 3466133557 final
[P3-1] Table 5 took 38.9567 sec, wrote 3532943585 right entries
[P3-2] Table 5 took 29.6419 sec, wrote 3532943585 left entries, 3532943585 final
[P3-1] Table 6 took 42.9072 sec, wrote 3713518554 right entries
[P3-2] Table 6 took 47.2412 sec, wrote 3713518554 left entries, 3713518554 final
[P3-1] Table 7 took 37.3316 sec, wrote 4294716442 right entries
[P3-2] Table 7 took 37.5831 sec, wrote 4294716442 left entries, 4294716442 final
Phase 3 took 436.531 sec, wrote 21876561194 entries to final plot
[P4] Starting to write C1 and C3 tables
[P4] Finished writing C1 and C3 tables
[P4] Writing C2 table
[P4] Finished writing C2 table
Phase 4 took 33.6745 sec, final plot size is 108832020311 bytes
Total plot creation time was 1807.85 sec

**Reply by marctanw90** [reply]
your above results are with a 5900x (24 thread) & 126GB ram setup? 110 Ram disk + excess 16GB enough to run the setup?

**Reply by jacek900** [reply]
This is correct.

**Reply by adamsky-01** [reply]
Jacek madMAx43v3r mentioned in the post below that once you use 104 GiB of 128 GiB RAM for minimum size of RAMdisk you might be able to utilize 12 maybe 16 threads only due to low remained memory. Did you notice any issues running 24 threads or did you use any trick to to get around it? 
https://github.com/madMAx43v3r/chia-plotter/discussions/150#discussioncomment-854256

**Reply by marctanw90** [reply]
ram disk 104 is pushing it. I reckon go 110 and leave the rest for system / buckets

**Reply by jacek900** [reply]
Running on all 24 threads and haven't noticed any issues so far. 128GB total system ram and 110GB ram disk.

**Reply by vctrb** [reply]
My 3900X does lab runs in 1600 seconds, production runs in 1660 seconds. 

Your nvme is not made for this job.

**Comment by casa10** [comment]
Some returns about CPU temperatures according each phases?

**Reply by adamsky-01** [reply]
![image](https://user-images.githubusercontent.com/60489596/121806973-06e13200-cc4a-11eb-9ed7-54846c34f487.png)
On the left current temps, on the right max. recorded temps.
I didn't check temps for each phase.

**Reply by adamsky-01** [reply]
![image](https://user-images.githubusercontent.com/60489596/121807076-85d66a80-cc4a-11eb-8536-74b70e1ed5bb.png)
It's almost silent setup. Plenty large, low spinning fans, Noctua NH-D15 coolers. Temps. are acceptable, the thermal throttling starts in somewhere between 90c and 100c so still way below 90deg.

**Reply by casa10** [reply]
Thank you.
Temperature often more than 70°C. After normal because if you use all threads or even half, processor will burn to make all the work.
Fortunately lik you said less than max autorized.

**Comment by leonidbkh** [comment]
I made last plot in 1700s on my 5800x with 110gb ramdisk and 1tb 970 evo plus as first temp. Still have some problems with CPU cooling, need to change cooling system, have 90 degrees at 1st phase
